src/exoticatechnologies/modifications/exotics/impl/RewindSystem.kt

```diff
@@ -391,33 +647,86 @@ class RewindSystem(key: String, settings: JSONObject) : Exotic(key, settings) {
             val hitpoints: Float = ship.hitpoints
-            val usableWeapons: List<WeaponAPI> = ship.usableWeapons.toList()
+            val weaponsData: List<WeaponData> = ship.allWeapons.toList().map { weapon -> WeaponData(weapon) }
+            val systemAmmo = ship.system?.let {it.ammo} ?: INVALID_INT_VALUE
+            val systemCooldownRemaining = ship.system?.let { it.cooldownRemaining } ?: INVALID_FLOAT_VALUE
+            val wasShieldOn: Boolean = ship.shield?.let { it.isOn } ?: false
+        }
+
+        data class MultiModuleShipParams(val ship: ShipAPI, override val timestamp: Float) : ActualShipParams() {
+            override val location: Vector2f = Vector2f(ship.location)
+            // now, since we have multiple modules, we will need to keep a track of all of them
+            val moduleParamsMap = HashMap<String, ConcreteShipParams>()
+
+            init {
+                val shipSections = getAllShipSections(ship)
+                shipSections.forEach { section ->
+                    moduleParamsMap[section.id] = ConcreteShipParams(section, timestamp)
+
+                    moduleParamsMap.toMap()
+                }
+            }
         }
 
-        object EmptyShipParams : ShipParams()
+        object EmptyShipParams : ActualShipParams() {
+            override val location: Vector2f
+                get() = throw IllegalStateException("EmptyShipParams does not have 'location'")
+            override val timestamp: Float
+                get() = throw IllegalStateException("EmptyShipParams does not have 'timestamp'")
+        }
+    }
 
+    data class WeaponData(val weapon: WeaponAPI) {
+        val currentHealth: Float = weapon.currHealth
+        val currentAngle: Float = weapon.currAngle
+        val currentAmmo: Int = weapon.ammo
+        val cooldownRemaining: Float = weapon.cooldownRemaining
+        val isDisabled: Boolean = weapon.isDisabled
+        val isPermanentlyDisabled: Boolean = weapon.isPermanentlyDisabled
     }
 
     companion object {
+        // To turn off logtag-based logging, make sure this list is empty
+        // Possible values are "ActivationCandidate", "Teleport"
+        private val DEBUG_LOGTAG_LIST: List<String> = listOf()
         private const val DEBUG = false
+
         private const val ITEM = "et_rewindchip"
         private const val MAX_TIME = 15f
         private const val COOLDOWN = 300
-        private const val PHASE_IN_DURATION = 2f;
+        private const val PHASE_IN_DURATION = 2f
+
+        private const val SYSTEM_ACTIVATION_SOUND = "gigacannon_charge"
+        private const val SHIP_TELEPORTED_SOUND = "tachyon_lance_fire"
+
+        private const val POST_TELEPORT_AFTERIMAGE_DURATION = 1f
+        private const val POST_TELEPORT_MAX_SHADOW_DISTANCE = 2500f
 
-        private val SYSTEM_ACTIVATION_SOUND = "gigacannon_charge"
-        private val SHIP_TELEPORTED_SOUND = "tachyon_lance_fire"
+        private const val INVALID_INT_VALUE = -5
```

Comment:
why is this -5 ? what kind of 'magic numbers' crap is this?

Reply:
it's just a "flag" value used here
```
            val systemAmmo = ship.system?.let {it.ammo} ?: INVALID_INT_VALUE
            val systemCooldownRemaining = ship.system?.let { it.cooldownRemaining } ?: INVALID_FLOAT_VALUE
```

so it needs to be int value, could have used Integer.MIN_VALUE i suppose but this one works just as fine too...

Reply:
resolving